### PR TITLE
Update smart contract parsing logic

### LINKED ISSUE #1237: Parser missing fileID entity persistence logic on CONTRACTCREATEINSTANCE transactions
**Detailed Description**
The mirror node currently persists CONTRACTCREATEINSTANCE transactions with eh assumption that fileId entities are already created on the network.
This is a valid assumption for the network but given the mirror node ability to start at any time the entity may be missing from the mirror db.
As such we should add logic to persist the file entity for CONTRACTCREATEINSTANCE which is found in the transaction body.

**Actual Behavior**
1. Submit a FILECREATE transaction to network for use by a contract
2. Start the mirror node parsing at a time after step 1
3. Submit a CONTRACTCREATEINSTANCE transaction to network
4. Mirror node would persist contract entity but would not at entity for file id already present on network

**Expected Behavior**
Mirror node would persist contract file entity in db

**Environment:**
- Version: v0.22.0-rc1

**Additional Context**
Additionally we are missing coverage for CONTRACTUPDATEINSTANCE, CONTRACTDELETEINSTANCE

### LINKED ISSUE #1664: Not persisting child contracts
**Detailed Description**
When a contract is created on Hedera the contract itself can contain the `new` operator to create additional contracts. These child contracts don't show up in the record stream as separate transactions. Until https://github.com/swirlds/services-hedera/issues/1947 was fixed in Mar. 2020, these child contract IDs weren't visible at all and did not inherit the properties of the parent. Now the list of child contract IDs will show up in the `TransactionRecord`.

On the mirror node, they don't get created at all as part of the parent contract create transaction despite them existing in main node state as separate accounts/contracts. Later, when that child contract is executed or otherwise referenced in a transaction we automatically create a entity row with almost no information associated with it.

**Actual Behavior**
1. Create a contract with the `new` operator
2. Inspect the `t_entities` table

**Expected Behavior**
The child contracts should get separate contract entities created and inherit the properties of the parent entity. 

**Environment:**
- Java: OpenJDK 11.0
- OS: Ubuntu 18.04
- Version: v0.27.0

**Additional Context**

### LINKED ISSUE #2632: Smart contracts 2.0 parsing
### Problem

We don't store detailed information about smart contracts in the database.

### Solution

Update the importer as noted in the design doc

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by steven-sheehy**
**Description**:
- Add `contract.parent_id` to track contracts created via the `new` operator
- Change parser to normalize contract result fields and persist contract logs
- Change `TransactionHandlerFactory` to be dynamic
- Change `TransactionHandler` to handle entity creation and notification
- Fix invalid `contract_result.amount` for contract calls due to previous migration
- Fix not persisting child contracts on contract create and contract call
- Fix not persisting file entity for contract transactions

**Related issue(s)**:

Fixes #1237
Fixes #1664
Fixes #2632

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListener.java [hunk 1]
```diff
@@ -317,13 +345,12 @@ public void onCustomFee(CustomFee customFee) throws ImporterException {
 
     @Override
     public void onEntity(Entity entity) throws ImporterException {
-        EntityId entityId = entity.toEntityId();
-        // handle empty id case
-        if (EntityId.isEmpty(entityId)) {
+        long id = entity.getId();
+        if (id == EntityId.EMPTY.getId() || contracts.containsKey(id)) {
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListener.java:349]: Do we reliably know the Entity type? If so another way we could do it is below instead
```suggestion
        if (id == EntityId.EMPTY.getId() || entity.getType() == EntityTypeEnum.CONTRACT.getId()) {
```

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListener.java [hunk 2]
```diff
@@ -424,7 +503,10 @@ private Entity mergeEntity(Entity cachedEntity, Entity newEntity) {
 
         if (newEntity.getKey() != null) {
             cachedEntity.setKey(newEntity.getKey());
-            cachedEntity.setPublicKey(newEntity.getPublicKey());
+        }
+
+        if (newEntity.getMaxAutomaticTokenAssociations() != null) {
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListener.java:508]: Remove as duplicate of below line

Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListener.java:508]: Merge issue. Removed locally now.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListener.java (hunk 1 above)
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListener.java:349]: We definitely don't know the type reliably. A crypto transfer for an contract could be sent to the `EntityListener` first and will show up as an account.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/ContractResultMigration.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                contractResult.getContractId(),
+                contractResult.getCreatedContractIds(),
+                contractResult.getErrorMessage(),
+                contractResult.getConsensusTimestamp()
+        );
+    }
+
+    private void insert(MigrationContractLog contractLog) {
+        jdbcTemplate.update("insert into contract_log (bloom, consensus_timestamp, contract_id, data, index, topic0, " +
+                        "topic1, topic2, topic3) values (?, ?, ?, ?, ?, ?)",
```
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/ContractResultMigration.java]: Yeah, actually fixed that right before your comments but you might've already reviewed that file by then.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/ContractLog.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    @JsonIgnore
+    @Override
+    public boolean isNew() {
+        return true;
+    }
+
+    @Data
+    public static class Id implements Serializable {
+        private long consensusTimestamp;
+        private int index;
```
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/ContractLog.java:85]: I'm not sure what you mean by inner transaction. Right now there's one contract_result per transaction and many contract_logs per single contract_result. So there's no need for contractid.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/AbstractEntity.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+@MappedSuperclass
+@NoArgsConstructor
+@SuperBuilder
+@TypeDef(
+        defaultForType = Range.class,
+        typeClass = PostgreSQLGuavaRangeType.class
+)
+public abstract class AbstractEntity {
+
+    private Long autoRenewPeriod;
```
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/AbstractEntity.java:52]: A contract does not have separate auto_renew_account_id as it has an implicit account that mirrors its contract id.

**Pull request merged**